File: android/core/src/main/java/app/visly/vml/viewimpl/TextViewImpl.kt
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package app.visly.vml.viewimpl

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.util.TypedValue
import android.widget.TextView
import app.visly.vml.*
import com.facebook.fbui.textlayoutbuilder.TextLayoutBuilder
import com.facebook.fbui.textlayoutbuilder.util.LayoutMeasureUtil

class TextViewImpl(ctx: Context): BaseViewImpl<TextView>(ctx) {
    internal var textAlign: Int = TextView.TEXT_ALIGNMENT_TEXT_START
    internal var maxLines: Int = Integer.MAX_VALUE
    internal var spacingMultiplier: Float = 1.0f
    internal var textSpan: Spannable = SpannableString("")

    override fun measure(width: Float?, height: Float?): Size {
        val textMeasureMode = when (width) {
            null -> TextLayoutBuilder.MEASURE_MODE_UNSPECIFIED
            else -> TextLayoutBuilder.MEASURE_MODE_EXACTLY
        }

        val textLayout = TextLayoutBuilder()
                .setText(textSpan)
                .setTextSize(Math.round(ctx.resources.displayMetrics.scaledDensity * 12f))
                .setWidth(width?.toInt() ?: Int.MAX_VALUE, textMeasureMode)
                .setIncludeFontPadding(true)
                .build()

        return if (textLayout != null) {
            Size(
                    width ?: textLayout.width.toFloat(),
                    height ?: LayoutMeasureUtil.getHeight(textLayout).toFloat())
        } else {
            Size(width ?: 0f, height ?: 0f)
        }
    }

    override fun setProp(key: String, value: JsonValue) {
        super.setProp(key, value)

        when (key) {
            "text-align" -> {
                textAlign = when (value) {
                    JsonValue.String("start") -> TextView.TEXT_ALIGNMENT_TEXT_START
                    JsonValue.String("end") -> TextView.TEXT_ALIGNMENT_TEXT_END
                    JsonValue.String("center") -> TextView.TEXT_ALIGNMENT_CENTER
                    else -> TextView.TEXT_ALIGNMENT_TEXT_START
                }
            }

            "max-lines" -> {
                maxLines = when (value) {
                    is JsonValue.Number -> value.value.toInt()
                    else -> Integer.MAX_VALUE
                }
            }

            "line-height" -> {
                spacingMultiplier = when (value) {
                    is JsonValue.Object -> (value.value["value"] as JsonValue.Number).value
                    else -> 1f
                }
            }

            "span" -> {
                textSpan = when (value) {
                    is JsonValue.Object -> makeSpan(value.value)
                    else -> SpannableString("")
                }
            }
        }
    }

    fun makeSpan(props: Map<String, JsonValue>): Spannable {
        val span = SpannableStringBuilder()

        when (val value = props["text"]) {
            is JsonValue.String -> { span.append(value.value) }
            is JsonValue.Array -> {
                for (i in 0 until value.value.size) {
                    when (val child = value.value.get(i)) {
                        is JsonValue.Object -> span.append(makeSpan(child.value))
                    }
                }
            }
        }

        val fontColor = when (val value = props["font-color"]) {
            is JsonValue.String -> parseColor(value.value)
            else -> null
        }

        val fontFamily = when (val value = props["font-family"]) {
            is JsonValue.String -> value.value
            else -> null
        }

        val fontStyle = when (val value = props["font-style"]) {
            JsonValue.String("normal") -> Typeface.NORMAL
            JsonValue.String("italic") -> Typeface.ITALIC
            else -> null
        }

        val fontWeight = when (val value = props["font-weight"]) {
            JsonValue.String("regular") -> Typeface.NORMAL
            JsonValue.String("bold") -> Typeface.BOLD
            else -> null
        }

        val fontSize = when (val value = props["font-size"]) {
            is JsonValue.Object -> value.toSips(ctx)
            else -> null
        }

        val start = 0
        val end = span.length

        span.setSpan(
                VMLTextSpan(fontStyle, fontWeight, fontSize, fontColor, fontFamily, ctx),
                start,
                end,
                Spanned.SPAN_INCLUSIVE_EXCLUSIVE)

        return span
    }

    override fun createView(): TextView {
        return TextView(ctx)
    }

    override fun bindView(view: TextView) {
        super.bindView(view)
        view.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
        view.setTextColor(Color.BLACK)
        view.setLineSpacing(0f, spacingMultiplier)
        view.maxLines = maxLines
        view.textAlignment = textAlign
        view.text = textSpan
    }
}

File: ios/VMLKit/Classes/TextViewImpl.swift
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit

private let systemFont = UIFont.systemFont(ofSize: 12);

internal class TextViewImpl: BaseViewImpl {
    internal var text: NSAttributedString = NSAttributedString()
    internal var numberOfLines: Int = -1
    internal var textAlignment: NSTextAlignment = .left
    internal var lineHeightMultiple = Float(1)
    
    override func measure(width: CGFloat?, height: CGFloat?) -> CGSize {
        let constraint = CGSize(width: width ?? CGFloat.greatestFiniteMagnitude, height: height ?? CGFloat.greatestFiniteMagnitude)
        let size = textWithLineHeight().boundingRect(with: constraint, options: [.usesLineFragmentOrigin], context: nil).size
        return CGSize(width: width ?? ceil(size.width), height: height ?? ceil(size.height))
    }
    
    override func setProp(key: String, value: JsonValue) {
        super.setProp(key: key, value: value)
        
        switch key {
        case "span": self.text = try! attributedString(from: try! value.asObject())
        case "max-lines": self.numberOfLines = Int(try! value.asNumber())
        case "line-height": self.lineHeightMultiple = try! value.asObject()["value"]!.asNumber()
        case "text-align":
            switch value {
            case .String(let value) where value == "start": self.textAlignment = .left
            case .String(let value) where value == "center": self.textAlignment = .center
            case .String(let value) where value == "end": self.textAlignment = .right
            default: self.textAlignment = .left
            }
        default: ()
        }
    }
    
    override func createView() -> UIView {
        let label = UILabel()
        label.textColor = .black
        label.font = UIFont.systemFont(ofSize: 12)
        return label
    }
    
    override func bindView(_ view: UIView) {
        super.bindView(view)
        
        let view = view as! UILabel
        view.attributedText = textWithLineHeight()
        view.textAlignment = self.textAlignment
        view.numberOfLines = self.numberOfLines
    }
    
    func attributedString(from props: [String: JsonValue]) throws -> NSAttributedString {
        let string: NSMutableAttributedString = try props.get("text") {
            switch $0 {
            case .String(let value):
                return NSMutableAttributedString(string: value)
            case .Array(let value):
                let parts = try value.map({ try attributedString(from: $0.asObject()) })
                return NSMutableAttributedString(attributedString: parts.reduce(NSAttributedString(), { $0 + $1 }))
            case let value: throw "Unexpected value for text: \(value)"
            }
        }
        
        let family: String = try props.get("font-family") {
            switch $0 {
            case .String(let value): return value
            case .Null: return systemFont.familyName
            case let value: throw "Unexpected value for font-family: \(value)"
            }
        }
        
        let italic: Bool = try props.get("font-style") {
            switch $0 {
            case .String(let value) where value == "normal": return false
            case .String(let value) where value == "italic": return true
            case .Null: return false
            case let value: throw "Unexpected value for font-style: \(value)"
            }
        }
        
        let size: CGFloat? = try props.get("font-size") {
            switch $0 {
            case .Object(let value): return CGFloat(try value.asDimension())
            case .Null: return nil
            case let value: throw "Unexpected value for font-size: \(value)"
            }
        }
        
        let weight: UIFont.Weight? = try props.get("font-weight") {
            switch $0 {
            case .String(let value) where value == "regular": return UIFont.Weight.regular
            case .String(let value) where value == "bold": return UIFont.Weight.bold
            case .Null: return nil
            case let value: throw "Unexpected value for font-weight: \(value)"
            }
        }
        
        if size != nil || weight != nil || italic {
            let descriptor = UIFontDescriptor(fontAttributes: [
                .family: family,
                .traits: [
                    UIFontDescriptor.TraitKey.weight: weight ?? UIFont.Weight.regular,
                    UIFontDescriptor.TraitKey.slant: italic ? 1 : 0,
                ]
                ])
            
            string.addAttribute(
                .font,
                value: UIFont(descriptor: descriptor, size: size ?? 12),
                range: NSRange(location: 0, length: string.length))
        }
        
        try props.get("font-color") {
            switch $0 {
            case .Null: ()
            case let value:
                string.addAttribute(
                    .foregroundColor,
                    value: try value.asColor().default,
                    range: NSRange(location: 0, length: string.length))
            }
        }
        
        return string
    }
    
    private func textWithLineHeight() -> NSAttributedString {
        let string = NSMutableAttributedString(attributedString: text)
        let paragraphStyle = NSMutableParagraphStyle()
        paragraphStyle.lineBreakMode = .byWordWrapping
        paragraphStyle.lineHeightMultiple = CGFloat(lineHeightMultiple)
        
        string.addAttribute(
            .paragraphStyle,
            value: paragraphStyle,
            range: NSRange(location: 0, length: string.length))
        
        return string
    }
}
